Editorial review: Document Translator and Language Detector APIs (#39457)
mdn/content: files/en-us/web/api/translator_and_language_detector_apis/index.md

---
title: Translator and Language Detector APIs
slug: Web/API/Translator_and_Language_Detector_APIs
page-type: web-api-overview
status:
  - experimental
browser-compat: api.Translator
spec-urls: https://webmachinelearning.github.io/translation-api
---

{{SeeCompatTable}}{{DefaultAPISidebar("Translator and Language Detector APIs")}}

> The **Translator and Language Detector APIs** provide functionality to detect the language a body of text is written in and translate text into different languages via a browser's own internal AI model.

Author: I've basically used this, but I've not swapped the order of the APIs in the title, as that's the actual name of the spec.

## Concepts and usage

Translating a body of text is a common task on today's web. Typical use cases include:

> - An on-the-fly translation of an article that isn't available in your language.

Reviewer: ```suggestion
- On-the-fly translation of an article that isn't available in your language.
```

> Translating a body of text is a common task on today's web, and one that AI is well-suited to. Typical use cases include:
> 
> - An on-the-fly translation of an article that isn't available in your language.
> - Translating a user's support requests into a language the support agent understands.
> - Facilitating chats between users that don't speak each other's languages.
> 
> For successful automated translation, it is vital to be able to detect the language of a body of text, however, language detection also has other uses beyond translation. You could automatically switch the language of UI strings or error/feedback messages based on the text input by a user. You could also detect the language of entered text and then automatically load the correct dictionary for spell checking or curse word detection.

Reviewer: I know AI will be used for this, but it isn't a requirement of the API, and this kind of feels like it is purely about AI stuff.
The segue also felt a bit unwieldy. Maybe something like:

```suggestion
Translating a body of text is a common task on today's web.
Typical use cases include:

- An on-the-fly translation of an article that isn't available in your language.
- Translating support requests into language that a support agent (human or AI) understands.
- Facilitating chats between users that don't speak each other's languages.

Detecting the language of a body of test is an important precursor for successful automated translation, but has other uses beyond direct translation.
For example, it allows automatic UI configuration based on user text entry, ranging from updating UI and error strings, to automatically loading appropriate dictionaries for spell checking or curse word detection.
```

Author: I've used some of your language here, but really this is about AI, so I've kept that focus a bit. The spec says "These APIs are part of a family of APIs expected to be powered by machine learning models..."

I think AI should be the focus until we get solid proof that a browser is powering this API using a non-AI mechanism. And I'm not sure what that would look like.

Reviewer: It would be beautiful :-)

> AI is well-suited to facilitating translation and language detection. The Translator and Language Detector APIs provide asynchronous ({{jsxref("Promise")}}-based) mechanisms for a website to detect languages and translate text via the browser's own internal AI model. This is useful and efficient because the browser handles the service, rather than the developer having to rely on the user downloading AI models, or host or pay for a cloud-based translation service.

Reviewer: ```suggestion
AI is well-suited to facilitating language detection and translation. The Translator and Language Detector APIs provide asynchronous ({{jsxref("Promise")}}-based) mechanisms for a website to detect languages and translate text via the browser's own internal AI model. This is useful and efficient because the browser handles the service, rather than the developer having to rely on the user downloading AI models, or host or pay for a cloud-based translation service.
```

- Language detection is done via the {{domxref("LanguageDetector")}} interface. A `LanguageDetector` object instance is created using the {{domxref("LanguageDetector.create_static", "LanguageDetector.create()")}} static method, then the {{domxref("LanguageDetector.detect", "detect()")}} instance method is passed the text string to detect the language for.
- Translation is done via the {{domxref("Translator")}} interface. A `Translator` object instance is created using the {{domxref("Translator.create_static", "Translator.create()")}} static method, then the {{domxref("Translator.translate", "translate()")}} instance method is passed the text string to translate.

> After the above instances have been created, you can remove them again using a `destroy()` instance method (for example, the {{domxref("Translator.destroy()")}}). You can also cancel pending operations using an {{domxref("AbortController")}}.

Reviewer: ```suggestion
The `Translator` and `LanguageDetector` instances consume a lot of resource, so once you're finished witt them, you are encouraged to remove them using a `destroy()` instance method (for example, the {{domxref("Translator.destroy()")}}). You can also cancel pending operations using an {{domxref("AbortController")}}.
```

Author: I've used this suggestion.

> See [Using the Translator and Language Detector APIs](/en-US/docs/Web/API/Translator_and_Language_Detector_APIs/Using) for a walkthrough of how the APIs work.

Reviewer: Maybe.
```suggestion
See [Using the Translator and Language Detector APIs](/en-US/docs/Web/API/Translator_and_Language_Detector_APIs/Using) for a walkthrough of how to use the APIs.
```

- {{httpheader("Permissions-Policy")}}; the {{httpheader("Permissions-Policy/translator", "translator")}} directive
  - : Controls access to the translation functionality. Where a policy specifically disallows its use, any attempts to call the `Translator` methods will fail with a `NotAllowedError` {{domxref("DOMException")}}.

## Security considerations

> The specification requires that a user has recently interacted with the page to create `LanguageDetector` and `Translator` objects ([transient user activation](/en-US/docs/Web/Security/User_activation) is required).
> 
> In addition, the specification controls access to the API via {{httpheader("Permissions-Policy/language-detector", "language-detector")}} and {{httpheader("Permissions-Policy/translator", "translator")}} {{httpheader("Permissions-Policy")}} directives.

Reviewer: The information is good. I don't usually reference the specification quite so directly, so as a suggestion

```suggestion
Creation of  `LanguageDetector` and `Translator` objects requires that the user has recently interacted with the page ([transient user activation](/en-US/docs/Web/Security/User_activation) is required).

Access to the API is also controlled via {{httpheader("Permissions-Policy/language-detector", "language-detector")}} and {{httpheader("Permissions-Policy/translator", "translator")}} {{httpheader("Permissions-Policy")}} directives.
```

Author: I copied that language directly from an existing MDN page. More pseudo-standards ;-)

I like your wording better, so I've used it here and will use it for future such instances.
